File: 14.最长公共前缀.js
/**
 * 也可以使用二分查找，二分查找的对象是前缀长度
 */

var longestCommonPrefix = function(strs){
  if (strs.length == 0) return "";
  if (strs.length == 1) return strs[0];

  let res = strs[0];
  for (let i = 1; i < strs.length; i++) {
      const count = Math.min(res.length, strs[i].length);
      let index = 0;
      while(index < count && res[index] == strs[i][index]) {
        index++;
      }
      res = res.substr(0, index);
  }

  return res;
};

File: 14.最长公共前缀.cpp
/*
	编写一个函数来查找字符串数组中的最长公共前缀。

	如果不存在公共前缀，返回空字符串 ""。


	示例 1：
	输入：strs = ["flower","flow","flight"]
	输出："fl"

	示例 2：
	输入：strs = ["dog","racecar","car"]
	输出：""
	解释：输入不存在公共前缀。
*/

class Solution {
public:
	string longestCommonPrefix(vector<string>& strs) {
		if (strs.size() == 0) return "";
		if (strs.size() == 1) return strs[0];

		string res = strs[0];
		for (int i = 1; i < strs.size(); i++) {
			int count = min(res.size(), strs[i].size());
			int index = 0;
			while (index < count && res[index] == strs[i][index]) {
				index++;
			}
			res = res.substr(0, index);
		}

		return res;
	}
};